File: glyph-mi/js/knowledge.js
/** Apply public knowledge packs to merged tag fields. */

function norm(s) {
  return String(s ?? '').trim().toLowerCase();
}

function basename(path) {
  const p = String(path || '').replace(/\\/g, '/');
  const i = p.lastIndexOf('/');
  return (i >= 0 ? p.slice(i + 1) : p).toLowerCase();
}

export function applyKnowledge(filePath, tags, fields, reasons, packs = []) {
  if (!packs.length) return { fields, reasons };

  const out = { ...fields };
  const rs = [...reasons];
  const base = basename(filePath);
  const rel = norm(filePath);

  for (const pack of packs) {
    for (const alias of pack.artistAliases || []) {
      const m = norm(alias.match);
      if (!m || !alias.artist) continue;
      const mode = alias.mode || 'tag';
      let hit = false;

      if (mode === 'path' || mode === 'folder') {
        hit = rel.includes(m) || rel.includes(`/${m}/`) || rel.includes(`\\${m}\\`);
      } else if (mode === 'tag') {
        hit = m === norm(tags?.artist);
      } else if (mode === 'basename') {
        hit = base === m || base.startsWith(`${m} -`) || base.startsWith(`${m}-`);
      } else {
        hit = m === norm(tags?.artist);
      }

      if (!hit) continue;
      const cur = norm(out.artist);
      if (!cur || cur === 'unknown artist' || (alias.force && mode !== 'tag')) {
        out.artist = alias.artist;
        rs.push(`knowledge: artist alias (${pack.id || 'pack'})`);
      }
    }

    for (const hint of pack.genreHints || []) {
      const pat = norm(hint.pattern);
      if (pat && (base.includes(pat) || rel.includes(pat)) && !out.genre && hint.genre) {
        out.genre = hint.genre;
        rs.push(`knowledge: genre hint (${pack.id || pack._sourceFile || 'pack'})`);
      }
    }

    for (const junk of pack.junkPatterns || []) {
      const pat = norm(junk.pattern);
      if (!pat) continue;
      if (base.includes(pat) || rel.includes(pat)) {
        rs.push(`knowledge: likely duplicate/junk filename (${pack.id || 'pack'})`);
      }
    }

    for (const rule of pack.folderRules || []) {
      const pat = norm(rule.match);
      if (pat && rel.includes(pat) && rule.genre && !out.genre) {
        out.genre = rule.genre;
        rs.push(`knowledge: folder rule (${pack.id || 'pack'})`);
      }
    }

    if (out.title) {
      for (const rule of pack.titleCleanup || []) {
        try {
          const re = new RegExp(rule.pattern, 'i');
          const next = out.title.replace(re, rule.replace ?? '');
          if (next !== out.title && next.trim()) {
            out.title = next.trim();
            rs.push('knowledge: title cleanup');
          }
        } catch {
          /* bad pattern */
        }
      }
    }

    for (const ex of pack.examples || []) {
      const refBase = norm(ex.ref?.basename);
      if (refBase && refBase === base) {
        const after = ex.after || ex.suggested || {};
        for (const key of ['title', 'artist', 'album', 'genre', 'year', 'trackNo']) {
          if (after[key] && !out[key]) out[key] = String(after[key]);
        }
        rs.push('knowledge pack match');
        break;
      }
    }
  }

  return { fields: out, reasons: rs };
}

File: glyph-mi/python/glyph_mi/knowledge.py
"""Load and apply public knowledge packs."""

from __future__ import annotations

import json
import os
from pathlib import Path

_PACKS: list[dict] | None = None


def glyph_mi_root() -> Path:
    return Path(__file__).resolve().parent.parent


def public_knowledge_dir() -> Path:
    return glyph_mi_root() / "knowledge" / "public"


def load_public_packs(force: bool = False) -> list[dict]:
    global _PACKS
    if _PACKS is not None and not force:
        return _PACKS

    packs: list[dict] = []
    kdir = public_knowledge_dir()
    if not kdir.is_dir():
        _PACKS = packs
        return packs

    for fp in sorted(kdir.glob("*.json")):
        try:
            data = json.loads(fp.read_text(encoding="utf-8"))
            data["_source_file"] = fp.name
            packs.append(data)
        except (json.JSONDecodeError, OSError):
            continue

    extra = os.environ.get("GLYPH_KNOWLEDGE_PATH", "")
    if extra:
        ep = Path(extra)
        if ep.is_file() and ep.suffix == ".json":
            try:
                packs.append(json.loads(ep.read_text(encoding="utf-8")))
            except (json.JSONDecodeError, OSError):
                pass
        elif ep.is_dir():
            for fp in ep.glob("*.json"):
                try:
                    packs.append(json.loads(fp.read_text(encoding="utf-8")))
                except (json.JSONDecodeError, OSError):
                    continue

    _PACKS = packs
    return packs


def _norm(s: str) -> str:
    return (s or "").strip().lower()


def _basename(path: str) -> str:
    return Path(path or "").name.lower()


def apply_knowledge(file_path: str, tags: dict, merged_fields: dict, reasons: list[str]) -> dict:
    """Enhance merged fields from public packs and examples."""
    packs = load_public_packs()
    if not packs:
        return merged_fields

    base = _basename(file_path)
    rel_hint = _norm(file_path.replace("\\", "/"))

    for pack in packs:
        for alias in pack.get("artistAliases") or []:
            if _norm(alias.get("match")) in _norm(tags.get("artist") or "") or _norm(alias.get("match")) in base:
                if alias.get("artist"):
                    merged_fields["artist"] = alias["artist"]
                    reasons.append(f"knowledge: artist alias ({pack.get('_source_file', 'pack')})")

        for hint in pack.get("genreHints") or []:
            pat = hint.get("pattern", "")
            if pat and (pat.lower() in base or pat.lower() in rel_hint):
                if hint.get("genre") and not merged_fields.get("genre"):
                    merged_fields["genre"] = hint["genre"]
                    reasons.append(f"knowledge: genre hint ({pack.get('_source_file', 'pack')})")

        for ex in pack.get("examples") or []:
            ref = ex.get("ref") or {}
            ref_base = _norm(ref.get("basename") or "")
            if ref_base and ref_base == _norm(base):
                after = ex.get("after") or ex.get("suggested") or {}
                for key in ("title", "artist", "album", "genre", "year", "trackNo"):
                    if after.get(key) and not merged_fields.get(key):
                        merged_fields[key] = str(after[key])
                reasons.append("knowledge pack match")
                break

    return merged_fields


def reload_knowledge() -> int:
    """Force reload packs; returns count loaded."""
    global _PACKS
    _PACKS = None
    return len(load_public_packs(force=True))
